projects/cli-files/go-cat/cmd/root.go

```diff
@@ -0,0 +1,108 @@
+package cmd
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"strings"
+)
+
+func Execute() {
+	// define a flag of type boolean with the specific properties (name, value, usage)
+	numberFlag := flag.Bool("n", false, "number all output lines")
+
+	// parse the command line flags
+	flag.Parse()
+
+	// if there is no argument given then error
+	if len(os.Args) < 2 {
+		fmt.Fprintf(os.Stderr, "no filename provided\n")
+		os.Exit(2)
+	}
+
+	// create a variable to store the file name
+	var filename string
+
+	if *numberFlag {
+		// if the -n flag is provided the filename will be the third value in the os.Args slice
+		filename = os.Args[2]
+	} else {
+		// if no flag is provided the filename will be the second value in the os.Args slice
+		filename = os.Args[1]
+	}
+
+	// try to get the file information
+	fileInfo, err := os.Stat(filename)
+
+	// if we cannot get the file information then error
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "could not read the file/directory information: %s\n", filename)
+		os.Exit(2)
+	}
+
+	// if the file is a directory then warn the user
+	if fileInfo.IsDir() {
+		fmt.Fprintf(os.Stderr, "%s : Is a directory\n", filename)
+		os.Exit(1)
+	}
+
+	// try to read the file
+	fileContents, err := os.ReadFile(filename)
```

Comment:
`os.ReadFile` will read the whole file into memory, which means e.g. if you have a 100GB file on a computer with 16GB RAM, your program will probably fail to allocate memory and crash (or if you have a 15GB file on a computer with 16GB RAM, you will find yourself without spare RAM for anything else). Can you avoid needing to read the whole file into memory at one time?

Reply:
Instead of reading the whole file I have tried instead to read line by line using a `bufio.NewReader`

> "bufio.NewReader wraps an io.Reader (such as a file) and provides buffering functionality. It reads a larger chunk of data from the underlying reader into an internal buffer. Subsequent reads are served from the buffer until it is exhausted, at which point the reader fetches more data from the underlying reader. This buffering mechanism can significantly improve reading performance, especially when dealing with many small reads or when reading from a slow input source like a network connection."

It gives me access to methods like ReadString which are abstractions that smart/talented people have written/tested
And those methods read from the buffer rather than making their own Read system calls(?)